src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs

```diff
@@ -92,59 +107,124 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
+                    throw new MongoAuthenticationException(connection.ConnectionId, "User not found.", ex);
+                }    
+                throw;
+            }
+        }
+
+        internal static IsMasterResult GetIsMasterResult(
+            IConnection connection,
+            CommandWireProtocol<BsonDocument> isMasterProtocol,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                return new IsMasterResult(isMasterProtocol.Execute(connection, cancellationToken));
+            }
+            catch (MongoCommandException ex)
+            {
+                // If the isMaster command fails with error code 11 (UserNotFound), drivers must consider authentication
+                // to have failed.In such a case, drivers MUST raise an error that is equivalent to what they would have
+                // raised if the authentication mechanism were specified and the server responded the same way.
+                if (ex.Code == 11)
+                {
+                    throw new MongoAuthenticationException(connection.ConnectionId, "User not found.", ex);
+                }
+                throw;
+            }
+        }
 
         // private methods
         private CommandWireProtocol<BsonDocument> CreateBuildInfoProtocol()
         {
             var buildInfoCommand = new BsonDocument("buildInfo", 1);
             var buildInfoProtocol = new CommandWireProtocol<BsonDocument>(
-                DatabaseNamespace.Admin,
-                buildInfoCommand,
-                true,
-                BsonDocumentSerializer.Instance,
-               null);
+                databaseNamespace: DatabaseNamespace.Admin,
+                command: buildInfoCommand,
+                slaveOk: true,
+                resultSerializer: BsonDocumentSerializer.Instance,
+                messageEncoderSettings: null);
             return buildInfoProtocol;
         }
 
         private CommandWireProtocol<BsonDocument> CreateGetLastErrorProtocol()
         {
             var getLastErrorCommand = new BsonDocument("getLastError", 1);
             var getLastErrorProtocol = new CommandWireProtocol<BsonDocument>(
-                DatabaseNamespace.Admin,
-                getLastErrorCommand,
-                true,
-                BsonDocumentSerializer.Instance,
-                null);
+                databaseNamespace: DatabaseNamespace.Admin,
+                command: getLastErrorCommand,
+                slaveOk: true,
+                resultSerializer: BsonDocumentSerializer.Instance,
+                messageEncoderSettings: null);
             return getLastErrorProtocol;
         }
 
-        internal BsonDocument CreateIsMasterCommand()
-        {
-            return CreateIsMasterCommand(_clientDocument);
+
+        internal static CommandWireProtocol<BsonDocument> CreateIsMasterProtocol(BsonDocument isMasterCommand)
+        {    
+            return MasterCommandToProtocol(isMasterCommand);
```

Comment:
Let's get rid of MasterCommandToProtocol and just move the body of MasterCommandToProcotol into here.

Reply:
Ack.